# * Kurapati Venkatesh

# * 09012021

# * Topic: Hypothesis Testing ✨

# ---------------------------------------------------------------------------

In [12]:
# import libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import scipy
from scipy import stats

## 1. A F&B manager wants to determine whether there is any significant difference in the diameter of the cutlet between two units. A randomly selected sample of cutlets was collected from both units and measured? Analyze the data and draw inferences at 5% significance level. Please state the assumptions and tests that you carried out to check validity of the assumptions. 

## ANSWER:

In [13]:
cut = pd.read_csv(r"G:\360digit\assignments meterial\Hypothesis testing\Cutlets.csv")

In [14]:
cut.head()

Unit A  Unit B
0  6.8090  6.7703
1  6.4376  7.5093
2  6.9157  6.7300
3  7.3012  6.7878
4  7.4488  7.1522

In [15]:
cu = cut.dropna()

In [16]:
cu.columns = ["A", "B"]

In [17]:
cu.head()

A       B
0  6.8090  6.7703
1  6.4376  7.5093
2  6.9157  6.7300
3  7.3012  6.7878
4  7.4488  7.1522

## Assumptions:

#### H0 : There is no significant differences between diameter of the cutlets between 2 units 
- Diameter of cutlets in A = Diameter of cutlets in B 


#### Ha : There is significant differences between diameter of the cutlets between 2 units 
- Diameter of cutlets in A >> Diameter of cutlets in B or Diameter of cutlets in A << Diameter of cutlets in B

Alpha = 0.05

#### Step 1: 
We need to determine wether the diameters are similar or not

#### Step 2: 
We need to determine which Test is to be done. 

Here y is continuous and x is discrete and there 2 independent variables, 

We now have to check if y1 and y2 are normal, for which we perform shapiro test

#### Step 3: 
Find Normality of the data.

##### Normality Test of Unit A:

* H0: Data are normal


* Ha: Data are not normal 


* if p-value is > 0.05 => Accept null hypothesis


* if p-value is < 0.05 =>Reject null hypothesis

In [26]:
stats.shapiro(cu.A)

ShapiroResult(statistic=0.9649458527565002, pvalue=0.3199819028377533)

##### p value = 0.32 > 0.05, therefor we accept null hypothesis, i.e Unit A is normal. 

In [22]:
stats.shapiro(cu.B)

ShapiroResult(statistic=0.9727300405502319, pvalue=0.5224985480308533)

###### p value = 0.5225 > 0.05, therefor we accept null hypothesis, i.e Unit A is normal

#### Step 4:
Now we have to check if variances are equal

In [23]:
scipy.stats.levene(cu.A, cu.B)

LeveneResult(statistic=0.6650897638632386, pvalue=0.4176162212502553)

###### p-value = 0.4176 > 0.05 therefore, variances are said to be equal. 

* So, we have to do 2- Sample t test. 

In [24]:
scipy.stats.ttest_ind(cu.A, cu.B)

Ttest_indResult(statistic=0.7228688704678063, pvalue=0.4722394724599501)

##### As observed above, p value > 0.05 and means are similar, therefore we can accept Null Hypothesis.

##### That is: 

##### H0: There is no significant differences between diameter of the cutlets between 2 units Diameter of cutlets in A = Diameter of cutlets in B

# ---------------------------------------------------------------------------------------------------

## 2. A hospital wants to determine whether there is any difference in the average Turn Around Time (TAT) of reports of the laboratories on their preferred list. They collected a random sample and recorded TAT for reports of 4 laboratories. TAT is defined as sample collected to report dispatch. Analyze the data and determine whether there is any difference in average TAT among the different laboratories at 5% significance level.

## ANSWER:

In [46]:
#libraries:
from scipy.stats import f_oneway
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [27]:
labtat = pd.read_csv("G:\\360digit\\assignments meterial\\Hypothesis testing\\lab_tat_updated.csv")

In [28]:
labtat.head()

Laboratory_1  Laboratory_2  Laboratory_3  Laboratory_4
0        185.35        165.53        176.70        166.13
1        170.49        185.91        198.45        160.79
2        192.77        194.92        201.23        185.18
3        177.33        183.00        199.61        176.42
4        193.41        169.57        204.63        152.60

In [29]:
labtat.isnull ().sum () 

Laboratory_1    0
Laboratory_2    0
Laboratory_3    0
Laboratory_4    0
dtype: int64

In [30]:
labtat.columns = ["l1", "l2", "l3", "l4"]

In [31]:
labtat.head()

l1      l2      l3      l4
0  185.35  165.53  176.70  166.13
1  170.49  185.91  198.45  160.79
2  192.77  194.92  201.23  185.18
3  177.33  183.00  199.61  176.42
4  193.41  169.57  204.63  152.60

## Assumptions:

##### H0 : There is no differences in the average TAT of 4 laboratories 

##### Ha : There is differences in the average TAT of 4 laboratories

##### Alpha = 0.05

### Step 1: 

We need to determine whether the average TAT of 4 laboratories are similar or not

### Step 2: 

We need to determine which Test is to be done.

- Here y is continuous and x is discrete and there 4 independent variables, 

- We now have to check if y1, y2, y3, y4 are normal, for which we perform shapiro test.

### Step 3: 

Find Normality of the data.

##### H0: Data are normal 
##### Ha: Data are not normal

##### if p-value is > 0.05 => Accept null hypothesis 
##### if p-value is < 0.05 =>Reject null hypothesis

In [47]:
stats.shapiro(labtat.l1)

ShapiroResult(statistic=0.9886691570281982, pvalue=0.42317795753479004)

- p-value = 0.4232 > 0.05 so p high null fly => It follows normal distribution

In [33]:
stats.shapiro(labtat.l2)

ShapiroResult(statistic=0.9936322569847107, pvalue=0.8637524843215942)

- p-value = 0.8637 > 0.05 so p high null fly => It follows normal distribution

In [34]:
stats.shapiro(labtat.l3)

ShapiroResult(statistic=0.9796067476272583, pvalue=0.06547004729509354)

- p-value = 0.06547 > 0.05 so p high null fly => It follows normal distribution

In [35]:
stats.shapiro(labtat.l4)

ShapiroResult(statistic=0.9913753271102905, pvalue=0.6618951559066772)

- p-value = 0.6619 > 0.05 so p high null fly => It follows normal distribution

From shapiro.test() we conclude that all the data are normal as p value is greater than 0.05.

### Step 4: 

Now we have to check if variances are equal

#### Statistical tests for comparing variances:

There are many solutions to test for the equality (homogeneity) of variance across groups, including:

###### F-test: 
- Compare the variances of two samples. The data must be normally distributed.


###### Bartlett’s test: 
- Compare the variances of k samples, where k can be more than two samples. The data must be normally distributed. The Levene test is an alternative to the Bartlett test that is less sensitive to departures from normality. 


###### Levene’s test: 
- Compare the variances of k samples, where k can be more than two samples. It’s an alternative to the Bartlett’s test that is less sensitive to departures from normality. 


###### Fligner-Killeen test: 
- a non-parametric test which is very robust against departures from normality.

I will perform The Levenet’s Test: 
- H0: variances are equal


- Ha: variances are unequal 


- if p-value is > 0.05 => Accept null hypothesis 


- if p-value is < 0.05 =>Reject null hypothesis

In [36]:
labtat.head()

l1      l2      l3      l4
0  185.35  165.53  176.70  166.13
1  170.49  185.91  198.45  160.79
2  192.77  194.92  201.23  185.18
3  177.33  183.00  199.61  176.42
4  193.41  169.57  204.63  152.60

In [37]:
scipy.stats.levene(labtat.l1, labtat.l2, labtat.l3, labtat.l4)

LeveneResult(statistic=1.025294593220823, pvalue=0.38107781677304564)

- From the output, it can be seen that the p-value of 0.5899 is greater than the significance level of 0.05. This means that there is no evidence to suggest that the variance in data is statistically significantly different for the three treatment groups.

Therefore, we perform one way ANOVA TEST

In [43]:
F, p = stats.f_oneway(labtat.l1, labtat.l2, labtat.l3, labtat.l4)

In [44]:
F

121.39264646442368

In [45]:
p

2.143740909435053e-58

- As the p-value is less than the significance level 0.05, we can conclude that there are significant differences between the groups highlighted with “*" in the model summary.

#### Multiple pairwise-comparison between the means of groups:

In one-way ANOVA test, a significant p-value indicates that some of the group means are different, but we don’t know which pairs of groups are different. 


It’s possible to perform multiple pairwise-comparison, to determine if the mean difference between specific pairs of group are statistically significant.

##### Tukey multiple pairwise-comparisons

As the ANOVA test is significant, we can compute Tukey HSD (Tukey Honest Significant Differences, R function: TukeyHSD()) for performing multiple pairwise-comparison between the means of groups. 


The function TukeyHD() takes the fitted ANOVA as an argument.

In [48]:
#perform one-way ANOVA
f_oneway(labtat.l1, labtat.l2, labtat.l3, labtat.l4)

F_onewayResult(statistic=121.39264646442368, pvalue=2.143740909435053e-58)

In [50]:
# Create four random groups of data with a mean difference of 1

mu, sigma = 10, 3 # mean and standard deviation
group1 = np.random.normal(mu, sigma, 50)

mu, sigma = 11, 3 # mean and standard deviation
group2 = np.random.normal(mu, sigma, 50)

mu, sigma = 12, 3 # mean and standard deviation
group3 = np.random.normal(mu, sigma, 50)

mu, sigma = 13, 3 # mean and standard deviation
group4 = np.random.normal(mu, sigma, 50)

# Show the results for Anova

F_statistic, pVal = stats.f_oneway(group1, group2, group3, group4)

print ('P value:')
print (pVal)

P value:
1.3879045556017055e-07


For the multicomparison tests we will put the data into a dataframe. And then reshape it to a stacked dataframe

In [51]:
# Put into dataframe
df = labtat
df = pd.DataFrame()
df['l1'] = group1
df['l2'] = group2
df['l3'] = group3
df['l4'] = group4

In [52]:
# Stack the data (and rename columns):

stacked_data = df.stack().reset_index()
stacked_data = stacked_data.rename(columns={'level_0': 'id',
                                            'level_1': 'lab',
                                            0:'result'})

In [53]:
# Show the first 8 rows:

print (stacked_data.head(8))

   id lab     result
0   0  l1  17.164282
1   0  l2   9.623453
2   0  l3   8.734313
3   0  l4  13.342110
4   1  l1   7.975878
5   1  l2  14.546414
6   1  l3  13.217575
7   1  l4  17.265002


In [55]:
from statsmodels.stats.multicomp import (pairwise_tukeyhsd,
                                         MultiComparison)

# Set up the data for comparison (creates a specialised object)
MultiComp = MultiComparison(stacked_data['result'],
                            stacked_data['lab'])

# Show all pair-wise comparisons:

# Print the comparisons

print(MultiComp.tukeyhsd().summary())

Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj   lower  upper  reject
---------------------------------------------------
    l1     l2   1.0606   0.28 -0.4723 2.5934  False
    l1     l3   1.6683  0.027  0.1354 3.2011   True
    l1     l4   3.5439  0.001   2.011 5.0767   True
    l2     l3   0.6077 0.7092 -0.9252 2.1406  False
    l2     l4   2.4833  0.001  0.9504 4.0162   True
    l3     l4   1.8756 0.0095  0.3427 3.4085   True
---------------------------------------------------


## Conclusion:

- Ha : There is differences in the average TAT of 4 laboratories except for Lab 1 and Lab 2

# -------------------------------------------------------------------------------------

## 3. Sales of products in four different regions is tabulated for males and females. Find if male-female buyer rations are similar across regions.
## H0 : All Proportions are equal
## Ha : Not All proportions are equal

## ANSWER:

In [57]:
BuyerRatio = pd.read_csv("G:\\360digit\\assignments meterial\\Hypothesis testing\\BuyerRatio.csv")

In [63]:
BuyerRatio

Observed Values  East  West  North  South
0           Males    50   142    131     70
1         Females   435  1523   1356    750

In [64]:
BuyerRatio.columns = ["gender", "E", "W", "N", "S"]

In [65]:
BuyerRatio.head()

gender    E     W     N    S
0    Males   50   142   131   70
1  Females  435  1523  1356  750

- Inputs are 4 discrete variables(east,west,north,south). 


- Output is also discrete. We are trying to find out if proportions of male and female are similar or not across the regions 


- We proceed with chi-square test

#### Create hypothesis: 
###### Ho= Proportions of Male and Female are same 
###### Ha= Proportions of Male and Female are not same 
###### Alpha=0.05

In [39]:
Chi_square


	Pearson's Chi-squared test

data:  BR
X-squared = 1.5959, df = 3, p-value = 0.6603


- if p-value is > 0.05 => Accept null hypothesis

## Conclusion:

- Ho= Proportions of Male and Female are same

# -------------------------------------------------------------------------------------

## 4. Telecall uses 4 centers around the globe to process customer order forms. They audit a certain % of the customer order forms. Any error in order form renders it defective and must be reworked before processing. The manager wants to check whether the defective % varies by center. Please analyze the data at 5% significance level and help the manager draw appropriate inferences

## ANSWER:

- Ho= There is no Variations in percentage of defect across different centres 
- Ha: There is Variations in percentage of defect across different centres
- Both x and y are discrete and there are more than 2 groups so we proceed with Chi squared Test 
+ Alpha=0.05

In [69]:
CustO = pd.read_csv(r"G:\360digit\assignments meterial\Hypothesis testing\CustomerOrderform.csv")

In [70]:
CustO.head()

Phillippines   Indonesia       Malta       India
0   Error Free  Error Free   Defective  Error Free
1   Error Free  Error Free  Error Free   Defective
2   Error Free   Defective   Defective  Error Free
3   Error Free  Error Free  Error Free  Error Free
4   Error Free  Error Free   Defective  Error Free

In [76]:
(CustO.Phillippines).describe()

count            300
unique             2
top       Error Free
freq             271
Name: Phillippines, dtype: object

In [77]:
CustO.Indonesia.describe()

count            300
unique             2
top       Error Free
freq             267
Name: Indonesia, dtype: object

In [78]:
CustO.Malta.describe()

count            300
unique             2
top       Error Free
freq             269
Name: Malta, dtype: object

In [79]:
CustO.India.describe()

count            300
unique             2
top       Error Free
freq             280
Name: India, dtype: object

In [10]:
CO

            directions
status       Phillipines Indonesia Malta India
  Defective           29        33    31    20
  Error Free         271       267   269   280

In [11]:
chisq_test(CO)


	Pearson's Chi-squared test

data:  CO
X-squared = 3.859, df = 3, p-value = 0.2771


- if p-value is > 0.05 => Accept null hypothesis

## Conclusion:

- Ho= There is no Variations in percentage of defect across different centres

# -----------------------------------------------------------------------------------------

## 5. Fantaloons Sales managers commented that % of males versus females walking into the store differ based on day of the week. Analyze the data and determine whether there is evidence at 5 % significance level to support this hypothesis.

## ANSWER:

######  Ho:  
* that % of males versus females walking in to the store DOES NOT differ based on day of the week


###### Ha: 
* that % of males versus females walking in to the store differ based on day of the week


###### Both x and y are discrete and there are 2 groups so we proceed with 2 ProportionTest 


###### Alpha: 0.05

In [80]:
Faltoons = pd.read_csv(r"G:\360digit\assignments meterial\Hypothesis testing\Fantaloons.csv")

In [81]:
Faltoons.head()

Weekdays Weekend
0     Male  Female
1   Female    Male
2   Female    Male
3     Male  Female
4   Female  Female

In [101]:
str(Faltoons)

'    Weekdays Weekend\n0       Male  Female\n1     Female    Male\n2     Female    Male\n3       Male  Female\n4     Female  Female\n..       ...     ...\n420      NaN     NaN\n421      NaN     NaN\n422      NaN     NaN\n423      NaN     NaN\n424      NaN     NaN\n\n[425 rows x 2 columns]'

In [96]:
Faltoons.Weekdays.describe()

count        400
unique         2
top       Female
freq         287
Name: Weekdays, dtype: object

In [97]:
Faltoons.Weekend.describe()

count        400
unique         2
top       Female
freq         233
Name: Weekend, dtype: object

In [89]:
from statsmodels.stats.proportion import proportions_ztest

tab1 = Faltoons.Weekdays.value_counts()
tab1

Female    287
Male      113
Name: Weekdays, dtype: int64

In [88]:
tab2 = Faltoons.Weekend.value_counts()
tab2

Female    233
Male      167
Name: Weekend, dtype: int64

In [103]:
p = scipy.stats.ttest_ind(tab1, tab2)

Ttest_indResult(statistic=0.0, pvalue=1.0)

In [20]:
p


	2-sample test for equality of proportions without continuity
	correction

data:  c(287, 233) out of c(400, 400)
X-squared = 16.022, df = 1, p-value = 6.261e-05
alternative hypothesis: two.sided
95 percent confidence interval:
 0.06956189 0.20043811
sample estimates:
prop 1 prop 2 
0.7175 0.5825 


- if p-value is < 0.05 => Reject null hypothesis

- And we can see in the above output that p value is less than 0.05

## Conclusion:

- Ha: that % of males versus females walking in to the store differ based on day of the week